bebraven/platform
# Implement adding a new Module like Course Resource

### Description

**Task**: https://app.asana.com/0/1174274412967132/1199394675251953

This PR allows you to **Add New** for a `Rise360Module` like a `CourseResource` on a course's edit page. This does **not** create a version; it also does **not** attach anything to the course.

Also, the LTI assignment placement for Rise360Modules won't work anymore because we've changed the controller and views.

### Motivation and Context

This is part of moving adding modules to Canvas courses to our course management page.

Because `Rise360Module` and `CourseResource` both have content that is a Articulate Rise360 zipfile, I added the concern `Attachable` to handle the `#new` (for the form) and `#create` (create the ActiveRecord model with the ActiveStorage attachment) to share the controller code. I had to make a couple changes to the models to share the controller code:

- `Rise360Module`: add `name` with DB migration
- `CourseResource`: rename `course_resource_zipfile` to `rise360_zipfile` in the model

The shared form lives in `views/attachable/_form.html.erb` and is shared by both sets of views.

### Test

- Course resource and Rise360 module specs

- Manually test:

- Go to a course's edit page, click **Add new** under _Edit Modules_:
<img width="1792" alt="Screen Shot 2020-12-03 at 4 55 44 PM" src="https://user-images.githubusercontent.com/62911934/101108293-125fec00-3589-11eb-93d6-36d16100aeeb.png">

- Fill out the form and click **Save**
<img width="1792" alt="Screen Shot 2020-12-03 at 4 58 16 PM" src="https://user-images.githubusercontent.com/62911934/101108266-01af7600-3589-11eb-835e-909d3f81a7ae.png">

- You'll have to wait a long time because of the downloading and unzipping and reuploading, but eventually:
<img width="1792" alt="Screen Shot 2020-12-03 at 4 59 03 PM" src="https://user-images.githubusercontent.com/62911934/101108240-f3f9f080-3588-11eb-8be0-d332603b247a.png">

- There's nothing in the UI that lets you see the Rise360Modules get, so go to the Rails console and check `Rise360Module.last` to see something like:

```
  Rise360Module Load (1.1ms)  SELECT "rise360_modules".* FROM "rise360_modules" ORDER BY "rise360_modules"."id" DESC LIMIT $1  [["LIMIT", 1]]
=> #<Rise360Module id: 4, created_at: "2020-12-04 00:58:17", updated_at: "2020-12-04 00:58:46", activity_id: "http://CVRBwM4pr955T7w5VaRCUOFE9zfZXe-x_rise", quiz_questions: 3, name: "Test Module #2">
```

### Types of changes




### Checklist:
